dbms/src/Interpreters/Aggregator.cpp

```diff
@@ -781,77 +822,173 @@ ALWAYS_INLINE void Aggregator::executeImplBatch(
             }
-        }
-        else
-        {
-            /// If a new key is inserted, initialize the states of the aggregate functions, and possibly something related to the key.
-            if (emplace_result.isInserted())
-            {
-                /// exception-safety - if you can not allocate memory or create states, then destructors will not be called.
-                emplace_result.setMapped(nullptr);
 
-                aggregate_data = aggregates_pool->alignedAlloc(total_size_of_aggregate_states, align_aggregate_states);
-                createAggregateStates(aggregate_data);
+            if unlikely (!emplace_result_holder.has_value())
+            {
+                LOG_INFO(log, "HashTable resize throw ResizeException since the data is already marked for spill");
+                break;
+            }
 
-                emplace_result.setMapped(aggregate_data);
+            auto & emplace_result = emplace_result_holder.value();
+            if constexpr (only_lookup)
+            {
+                if constexpr (compute_agg_data)
+                {
+                    if (emplace_result.isFound())
+                    {
+                        aggregate_data = emplace_result.getMapped();
+                    }
+                    else
+                    {
+                        agg_process_info.not_found_rows.push_back(j);
+                    }
+                }
+                else
+                {
+                    if (!emplace_result.isFound())
+                        agg_process_info.not_found_rows.push_back(i);
+                }
             }
             else
             {
-                aggregate_data = emplace_result.getMapped();
+                if constexpr (compute_agg_data)
+                {
+                    if (emplace_result.isInserted())
+                    {
+                        // exception-safety - if you can not allocate memory or create states, then destructors will not be called.
+                        emplace_result.setMapped(nullptr);
+
+                        aggregate_data
+                            = aggregates_pool->alignedAlloc(total_size_of_aggregate_states, align_aggregate_states);
+                        createAggregateStates(aggregate_data);
+
+                        emplace_result.setMapped(aggregate_data);
+                    }
+                    else
+                    {
+                        aggregate_data = emplace_result.getMapped();
+
+                        if constexpr (collect_hit_rate)
+                            ++agg_process_info.hit_row_cnt;
 
-                if constexpr (collect_hit_rate)
-                    ++agg_process_info.hit_row_cnt;
+                        if constexpr (enable_prefetch)
+                            __builtin_prefetch(aggregate_data);
+                    }
+                }
+                else
+                {
+                    emplace_result.setMapped(place);
```

Comment:
Is it save to push the fake address?

Reply:
it's ok. no one use value if agg func is zero. Also check: https://github.com/ClickHouse/ClickHouse/blob/70945668f81a3621685edf4d21728efc1ae4a902/src/Interpreters/Aggregator.cpp#L1020